src/server/transaction.cc

```diff
@@ -890,42 +890,10 @@ void Transaction::ExecuteAsync() {
     return;
   }
 
-  // We verify seq lock has the same generation number. See below for more info.
-  auto cb = [seq, this] {
-    EngineShard* shard = EngineShard::tlocal();
-
-    bool is_armed = IsArmedInShard(shard->shard_id());
-    // First we check that this shard should run a callback by checking IsArmedInShard.
-    if (is_armed) {
-      uint32_t seq_after = seqlock_.load(memory_order_relaxed);
-
-      DVLOG(3) << "PollExecCb " << DebugId() << " sid(" << shard->shard_id() << ") "
-               << run_count_.load(memory_order_relaxed);
-
-      // We also make sure that for multi-operation transactions like Multi/Eval
-      // this callback runs on a correct operation. We want to avoid a situation
-      // where the first operation is executed and the second operation is armed and
-      // now this callback from the previous operation finally runs and calls PollExecution.
-      // It is usually ok, but for single shard operations we abuse index 0 in shard_data_
-      // Therefore we may end up with a situation where this old callback runs on shard 7,
-      // accessing shard_data_[0] that now represents shard 5 for the next operation.
-      // seqlock provides protection for that so each cb will only run on the operation it has
-      // been tasked with.
-      // We also must first check is_armed and only then seqlock. The first check ensures that
-      // the coordinator thread crossed
-      // "run_count_.store(unique_shard_cnt_, memory_order_release);" barrier and our seqlock_
-      // is valid.
-      if (seq_after == seq) {
-        // shard->PollExecution(this) does not necessarily execute this transaction.
-        // Therefore, everything that should be handled during the callback execution
-        // should go into RunInShard.
-        shard->PollExecution("exec_cb", this);
-      } else {
-        VLOG(1) << "Skipping PollExecution " << DebugId() << " sid(" << shard->shard_id() << ")";
-      }
-    }
+  auto cb = [this] {
+    EngineShard::tlocal()->PollExecution("exec_cb", this);
 
-    DVLOG(3) << "ptr_release " << DebugId() << " " << seq;
+    DVLOG(3) << "ptr_release " << DebugId();
     intrusive_ptr_release(this);  // against use_count_.fetch_add above.
   };
```

Comment:
I still don't understand why we need seqlock 🙂 

Our code is: 

```c++
void RunInShard() {
  CHECK(armed);
  armed = false;

  cb_();
}
```

We disarm atomically (in fiber sense). Now cb is stateless - if there is work, it's correct to do it. 

Maybe seqlock was introduced because:
1. RunQuickie() didn't disarm atomically before
2. We never use shard_data[0] for atomic multi transactions. Maybe we did before? Need to check

Reply:
```
 It is usually ok, but for single shard operations we abuse index 0 in shard_data_
      // Therefore we may end up with a situation where this old callback runs on shard 7,
      // accessing shard_data_[0] that now represents shard 5 for the next operation.
```

Reply:
I see that we did it for non atomic multi. See `if ((key_index.HasSingleKey() && !IsAtomicMulti()) || is_stub)` in InitByKeys

Reply:
Now, one year later, I do not know how much we can rely on usefulness of non-atomic multi/lua transactions.

Reply:
> I see that we do it for non atomic multi

Yes, I figured that out too.

1. Yet they always keep the full backing array, so they don't have to rely on shard_data[0]. It's a bug
2. Even if, the poll task is stateless, so as long as we unarm without fiber suspensions, there is nothing wrong in pulling them again

Reply:
the problem is with the following sequence `MULTI, GET shard1, GET shard2, EXEC`, if we reuse shard_data[0], then it could be that the late callback on shard1 will run and execute the callback for the armed GET operation for shard2, obviously won't find shard2 key there. Then shard2 callback will run but it will be unarmed of course. so it's about running on wrong threads due to the confusion with the data structures. we remove this complexity and then can get rid of seqlock as well